[PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55940 [PyTorch] Always update op.current_dtype in TensorIteratorBase::set_output
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* **#55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor**
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned

Upcoming changes to `MaybeOwned<T>` will require that T is
one of these two types and will have custom code for both.

This diff updates the tests to continue to build under these new
requirements; it is being sent separately to demonstrate that the
tests continue to work on the current implementation.

Differential Revision: [D27630289](https://our.internmc.facebook.com/intern/diff/D27630289/)

swolchok committed:
[PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor

Upcoming changes to `MaybeOwned<T>` will require that T is
one of these two types and will have custom code for both.

This diff updates the tests to continue to build under these new
requirements; it is being sent separately to demonstrate that the
tests continue to work on the current implementation.

Differential Revision: [D27630289](https://our.internmc.facebook.com/intern/diff/D27630289/)

[ghstack-poisoned]

swolchok committed:
rebase on "[PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor"


Upcoming changes to `MaybeOwned<T>` will require that T is
one of these two types and will have custom code for both.

This diff updates the tests to continue to build under these new
requirements; it is being sent separately to demonstrate that the
tests continue to work on the current implementation.

Differential Revision: [D27630289](https://our.internmc.facebook.com/intern/diff/D27630289/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/test/MaybeOwned_test.cpp:
```diff
@@ -0,0 +1,297 @@
+void assertBorrow(const MaybeOwned<T>&, const T&);
+
+template <typename T>
+void assertOwn(const MaybeOwned<T>&, const T&, size_t useCount = 2);
```
Personally I'd omit the default param here, I think it just adds indirection when reading the callsites, but not a strong opinion

swolchok committed:
Update on "[PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor"


Upcoming changes to `MaybeOwned<T>` will require that T is
one of these two types and will have custom code for both.

This diff updates the tests to continue to build under these new
requirements; it is being sent separately to demonstrate that the
tests continue to work on the current implementation.

Differential Revision: [D27630289](https://our.internmc.facebook.com/intern/diff/D27630289/)

[ghstack-poisoned]
